### PR TITLE
[Model] Add support for Aya-23 8B Model by Cohere

### PR DESCRIPTION
**Pull request by GunjanDhanuka**
This PR adds support for the Aya-23 8B model, whose weights and config can be found here: https://huggingface.co/CohereForAI/aya-23-8B/tree/main

Also fixed a typo where `LlamaForCausalLM` was written as `LlamaForCasualLM` 

There was an issue with CUDA graph while compiling the model, so use `--opt "flashinfer=1;cublas_gemm=1;cudagraph=0"` to the `mlc_llm compile` command as suggested by @MasterJH5574 .

Solves issue: https://github.com/mlc-ai/web-llm/issues/483

### CODE REVIEW

File: python/mlc_llm/interface/gen_config.py [hunk 1]
```diff
@@ -129,7 +129,7 @@ def gen_config(  # pylint: disable=too-many-locals,too-many-arguments,too-many-b
         prefill_chunk_size=model_config.prefill_chunk_size,
         attention_sink_size=getattr(model_config, "attention_sink_size", -1),
         tensor_parallel_shards=model_config.tensor_parallel_shards,
-        conv_template=conversation,
+        conv_template=conversation,  # type: ignore
```
Comment (MasterJH5574) [inline python/mlc_llm/interface/gen_config.py:132]: Just curious why we disable mypy for this line?

File: python/mlc_llm/model/cohere/cohere_loader.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                functools.partial(
+                    lambda x, dtype: x.astype(dtype),
+                    dtype=mlc_param.dtype,
+                ),
+            )
+
+    return mapping
+
+
+# def awq(model_config: CohereConfig, quantization: Quantization) -> ExternMapping:
```
Comment (MasterJH5574) [inline python/mlc_llm/model/cohere/cohere_loader.py:94]: Please remove if we don't support.

File: python/mlc_llm/interface/gen_config.py (hunk 1 above)
Comment (GunjanDhanuka) [reply python/mlc_llm/interface/gen_config.py:132]: PyLance was raising a `reportArgumentType` issue for this line, attached the screenshot below:

image

File: python/mlc_llm/model/cohere/cohere_loader.py (hunk 2 above)
Comment (GunjanDhanuka) [reply python/mlc_llm/model/cohere/cohere_loader.py:94]: Just checked that AWQ weights for Aya-23 are available: https://huggingface.co/alijawad07/aya-23-8B-AWQ-GEMM

In that case, we can support it, I will uncomment the awq part in `cohere_loader.py`

**Comment by alphaarea** [comment]
Does this PR support [CohereForAI/c4ai-command-r-plus](https://huggingface.co/CohereForAI/c4ai-command-r-plus) ? They're both `CohereForCausalLM`

**Comment by MasterJH5574** [comment]
Will look into the tokenizer issue during `mlc_llm chat`.

**Comment by Ubospica** [comment]
@GunjanDhanuka The tokenizer issue is solved in this PR: #2649. Please tell me if there are any other related problems!

**Comment by tqchen** [comment]
@GunjanDhanuka please rebase and cross check if things can run

**Comment by GunjanDhanuka** [comment]
> @GunjanDhanuka please rebase and cross check if things can run

Yes the tokenizer issue is now resolved, but there was a delay because of the discrepancy in outputs from `mlc_llm chat` and ` python -m mlc_llm.testing.debug_chat` .  The current commit (https://github.com/mlc-ai/mlc-llm/pull/2603/commits/599f14842f7aa37dbe8225fe83e2bdd06a3ebf14) works fine on the `debug_chat` but `mlc_llm chat` is a hit or miss at times.
cc: @MasterJH5574 @tqchen

File: 3rdparty/tokenizers-cpp
Comment (MasterJH5574) [inline 3rdparty/tokenizers-cpp:1]: Seems that the 3rdparty tokenizer repo is downgraded. Could you revert this change?

**Review by MasterJH5574 — CHANGES_REQUESTED** [review]

File: 3rdparty/tokenizers-cpp
Comment (GunjanDhanuka) [reply 3rdparty/tokenizers-cpp:1]: Reversed those changes!

**Review by MasterJH5574 — APPROVED** [review]
Thank you @GunjanDhanuka!

**Comment by DenisSergeevitch** [comment]
Hi, I tried to convert a new version of [aya-expanse-8b](https://huggingface.co/CohereForAI/aya-expanse-8b) using this command:
`mlc_llm compile "/content/dist/aya-expanse-8b-q4f16_1-MLC-webgpu/mlc-chat-config.json" --device webgpu -o /content/dist/aya-expanse-8b-q4f16_1-MLC-webgpu.wasm` and it looks like the converted model does not perform well due to some convertion issues, however no errors have been encountered: 
```
[2024-11-08 12:38:33] INFO auto_config.py:70: Found model configuration: /content/dist/aya-expanse-8b-q4f16_1-MLC-webgpu/mlc-chat-config.json
[2024-11-08 12:38:33] INFO auto_config.py:154: Found model type: cohere. Use `--model-type` to override.
Compiling with arguments:
  --config          CohereConfig(model_type='cohere', hidden_size=4096, vocab_size=256000, num_hidden_layers=32, num_attention_heads=32, num_key_value_heads=8, intermediate_size=14336, layer_norm_eps=1e-05, position_embedding_base=10000, context_window_size=8192, prefill_chunk_size=8192, head_dim=128, tensor_parallel_shards=1, max_batch_size=128, kwargs={})
  --quantization    GroupQuantize(name='q4f16_1', kind='group-quant', group_size=32, quantize_dtype='int4', storage_dtype='uint32', model_dtype='float16', linear_weight_layout='NK', quantize_embedding=True, quantize_final_fc=True, num_elem_per_storage=8, num_storage_per_group=4, max_int_value=7, tensor_parallel_shards=0)
  --model-type      cohere
  --target          {"host": {"kind": "llvm", "tag": "", "keys": ["cpu"], "mtriple": "wasm32-unknown-unknown-wasm"}, "max_num_threads": runtime.BoxInt(256), "kind": "webgpu", "tag": "", "keys": ["webgpu", "gpu"]}
  --opt             flashinfer=0;cublas_gemm=0;faster_transformer=0;cudagraph=0;cutlass=0;ipc_allreduce_strategy=NONE
  --system-lib-prefix ""
  --output          /content/dist/aya-expanse-8b-q4f16_1-MLC-webgpu.wasm
  --overrides       context_window_size=None;sliding_window_size=None;prefill_chunk_size=None;attention_sink_size=None;max_batch_size=None;tensor_parallel_shards=None;pipeline_parallel_stages=None
[2024-11-08 12:38:33] INFO compile.py:140: Creating model from: CohereConfig(model_type='cohere', hidden_size=4096, vocab_size=256000, num_hidden_layers=32, num_attention_heads=32, num_key_value_heads=8, intermediate_size=14336, layer_norm_eps=1e-05, position_embedding_base=10000, context_window_size=8192, prefill_chunk_size=8192, head_dim=128, tensor_parallel_shards=1, max_batch_size=128, kwargs={})
[2024-11-08 12:38:33] INFO compile.py:158: Exporting the model to TVM Unity compiler
[2024-11-08 12:38:36] INFO compile.py:164: Running optimizations using TVM Unity
[2024-11-08 12:38:36] INFO compile.py:185: Registering metadata: {'model_type': 'cohere', 'quantization': 'q4f16_1', 'context_window_size': 8192, 'sliding_window_size': -1, 'attention_sink_size': -1, 'prefill_chunk_size': 8192, 'tensor_parallel_shards': 1, 'pipeline_parallel_stages': 1, 'kv_state_kind': 'kv_cache', 'max_batch_size': 128}
[2024-11-08 12:38:36] WARNING auto_target.py:130: --system-lib-prefix is not specified when building a static library
[2024-11-08 12:38:38] INFO pipeline.py:54: Running TVM Relax graph-level optimizations
[2024-11-08 12:38:42] INFO pipeline.py:54: Lowering to TVM TIR kernels
[2024-11-08 12:38:47] INFO pipeline.py:54: Running TVM TIR-level optimizations
[2024-11-08 12:39:06] INFO pipeline.py:54: Running TVM Dlight low-level optimizations
[2024-11-08 12:39:08] INFO pipeline.py:54: Lowering to VM bytecode
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `alloc_embedding_tensor`: 64.00 MB
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `batch_decode`: 14.00 MB
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `batch_prefill`: 897.00 MB
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `batch_verify`: 896.00 MB
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `create_tir_paged_kv_cache`: 0.00 MB
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `decode`: 0.11 MB
[2024-11-08 12:39:11] INFO estimate_memory_usage.py:58: [Memory usage] Function `embed`: 64.00 MB
[2024-11-08 12:39:12] INFO estimate_memory_usage.py:58: [Memory usage] Function `prefill`: 896.01 MB
[2024-11-08 12:39:12] INFO estimate_memory_usage.py:58: [Memory usage] Function `softmax_with_temperature`: 0.00 MB
[2024-11-08 12:39:13] INFO pipeline.py:54: Compiling external modules
[2024-11-08 12:39:13] INFO pipeline.py:54: Compilation complete! Exporting to disk
[12:39:18] /workspace/tvm/src/target/llvm/codegen_llvm.cc:185: Warning: Set native vector bits to be 128 for wasm32
[2024-11-08 12:39:57] INFO model_metadata.py:95: Total memory usage without KV cache:: 5203.76 MB (Parameters: 4306.76 MB. Temporary buffer: 897.00 MB)
[2024-11-08 12:39:57] INFO model_metadata.py:103: To reduce memory usage, tweak `prefill_chunk_size`, `context_window_size` and `sliding_window_size`
[2024-11-08 12:39:57] INFO compile.py:207: Generated: /content/dist/aya-expanse-8b-q4f16_1-MLC-webgpu.wasm
```

Is this bug specific to the new version of Aya? 

Model:
https://huggingface.co/shirman/aya-expanse-8b-MLC-WEBGPU

WebLLM demo:
https://shir-man.com/totally-not-a-deepl-clone/

**Comment by DenisSergeevitch** [comment]
@MasterJH5574 hello, can you please help to guide how to debug the model re-compile in [this](https://github.com/mlc-ai/mlc-llm/pull/2603#issuecomment-2470475026) case?
